lib/galaxy/webapps/galaxy/services/history_contents.py

```diff
@@ -1600,9 +1610,30 @@ def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
             return self.dataset_collection_manager.delete(trans, "history", item.id, recursive=True, purge=True)
         self.hda_manager.purge(item, flush=self.flush)
 
-    def _change_datatype(self, item: HistoryItemModel, params: ChangeDatatypeOperationParams):
+    def _change_datatype(
+        self, item: HistoryItemModel, params: ChangeDatatypeOperationParams, trans: ProvidesHistoryContext
+    ):
         if isinstance(item, HistoryDatasetAssociation):
-            item.change_datatype(params.datatype)
+            self._ensure_can_change_hda_datatype(item)
+            self._set_hda_datatype(item, params.datatype, trans)
```

Comment:
```suggestion
            item.dataset.state = item.dataset.states.SETTING_METADATA
            self._set_hda_datatype(item, params.datatype, trans)
```
For large files this can still take a while, so we should switch to the SETTING_METADATA state here, which means the dataset is considered "pending".